lib/simp/cli/commands/bootstrap.rb

```diff
@@ -1,66 +1,335 @@
+    end
+
+    bootstrap_start_time = Time.now
+
+    # Set us up to use the SIMP environment, if this has not already been done.
+    # (1) Be careful to preserve the existing primary, 'production' environment,
+    #     if one exists.
+    # (2) Create links to production in both the primary and secondary environment
+    #     paths.
+    environment_path = ::Utils.puppet_info[:simp_environment_path]
+    fail("Could not find the simp environment path at #{environment_path}") unless File.directory?(environment_path)
+
+    item = Simp::Cli::Config::Item::SetProductionToSimpAction.new
+    item.start_time = bootstrap_start_time
+    item.apply
+    fail ("Could not set 'simp' to production environment") unless item.applied_status == :succeeded
+
+    linecounts = Array.new
+
+    # Open log file
+    logfilepath = File.expand_path(BOOTSTRAP_LOG)
+    FileUtils.mkpath(File.dirname(logfilepath)) unless File.exists?(logfilepath)
+    @logfile = File.open(logfilepath, 'w')
+
+    # Print intro
+    system('clear')
+    say "=== Starting SIMP Bootstrap ===".yellow.bold
+
+    # Set an interrupt trap if safe mode is enabled
+    say "> The log can be found at '#{@logfile.path}'\n"
+    if not @unsafe
+      signals = ["INT","HUP","USR1","USR2"]
+      signals.each do |sig|
+        Signal.trap(sig) { say "\nSafe mode enabled, ignoring interrupt".magenta }
+      end
+      say "> Interrupts will be captured and ignored to ensure bootstrap integrity.".magenta.bold
+    else
+      say "> WARNING: Any interrupts may cause system instability.".red.bold
+    end
+
+    # Determine if a puppet agent is running, and what to do with it.
+    # Note: Killing a puppet agent (that started via cron) will release the lock it has on
+    # /var/lock/puppetagent_cron
+    _kill_agent = false
+    if @kill_agent
+      _kill_agent = true
+    else
+      if File.exists?(::Utils.puppet_info[:config]['agent_catalog_run_lockfile'])
+        say "> Detected agent lock file, #{::Utils.puppet_info[:config]['agent_catalog_run_lockfile']}".magenta
+        # If kill agent is not true or false, prompt
+        if not @kill_agent == false
+          kill_ask = ask ("> Do you wish to remove the agent lock and forcibly kill currently running agents? (yes|no) ".yellow) { |q| q.validate = /(yes)|(no)/i}
+          _kill_agent = true if kill_ask.downcase == 'yes'
+        end
+        # If the user does not intend to forcibly kill running agents but an agent
+        # is running, warn them and don't continue.
+        if _kill_agent == false
+          say "> Not removing lock or killing active puppet agent process(es)".magenta
+          say "> You will need to resolve all locks and running agent processes before continuing".magenta
+          exit 1
+        end
+      else
+        say "> Did not detect agent lock file"
+      end
+    end
+    if _kill_agent
+      say "> Killing puppet agents".cyan
+      system("pkill -9 -f 'puppet agent' >& /dev/null")
```

Comment:
@lnemsick-simp Puppet likes to ignore SIGINT at times. I suppose we could SIGINT it, wait for some amount of time and then SIGKILL it.